workflows/prognostic_c48_run/runtime/steppers/reservoir.py

```diff
@@ -0,0 +1,233 @@
+        self._last_called = self.PREDICT
+
+    def set_init_time(self, time: cftime.DatetimeJulian):
+        self._init_time = time
+
+    def __call__(self, state: str):
+        if state == self.INCREMENT:
+            self.to_incremented()
+        elif state == self.PREDICT:
+            self.to_predicted()
+        else:
+            raise ValueError(
+                f"Unknown state provided to _ReservoirStepperState {state}"
+            )
+
+
+class _ReservoirStepper:
+
+    label = "hybrid_reservoir"
+
+    def __init__(
+        self,
+        model: ReservoirAdapter,
+        reservoir_timestep: timedelta,
+        synchronize_steps: int,
+        model_timestep_seconds: int = 900,
+        state_machine: Optional[_FiniteStateMachine] = None,
+    ):
+        self.model = model
+        self.rc_timestep = reservoir_timestep
+        self.synchronize_steps = synchronize_steps
+        self.dt_atmos = timedelta(seconds=model_timestep_seconds)
+
+        if state_machine is None:
+            self._state_machine = _FiniteStateMachine()
+        else:
+            self._state_machine = state_machine
+
+    @property
+    def init_time(self):
+        return self._state_machine.init_time
+
+    @property
+    def completed_sync_steps(self):
+        return self._state_machine.completed_increments
+
+    def increment_reservoir(self, time, state):
+        """Should be called at beginning of time loop"""
+
+        if self._state_machine.init_time is None:
+            self._state_machine.set_init_time(time)
+
+        if self._is_rc_update_step(time):
+            reservoir_inputs = state[self.model.input_variables]
+
+            try:
+                n_halo_points = self.model.rank_divider.overlap
+                rc_in_with_halos = append_halos_using_mpi(
+                    reservoir_inputs, n_halo_points
+                )
+            except RuntimeError:
+                raise ValueError(
+                    "MPI not available or tile dimension does not exist in state fields"
+                    " during reservoir increment update"
+                )
+
+            if self.completed_sync_steps == 0:
+                self.model.reset_state()
+
+            self._state_machine(self._state_machine.INCREMENT)
+            self.model.increment_state(rc_in_with_halos)
+
+        return {}, {}, {}
+
+    def predict(self, time, state):
+        """Called at the end of timeloop after time has ticked from t -> t+1"""
+
+        updated_state = {}
+        diags = {}
+
+        # won't evaluate to true until we've reached the step before the next increment
+        # e.g., if fv3 has k timesteps between rc timestep, on t + k - 1, the timestep
+        # at the end will have ticked over to t + k, we'll maybe use the integrated
+        # hybrid quantites from t -> t + k, make the rc prediction for t + k, and then
+        # increment during the next time loop based on those outputs.
+        if self._is_rc_update_step(time):
```

Comment:
Is it possible to move this check up a level to the increment/predict stepper classes' `__call__`? That way, `increment_state` and `predict` wouldn't need the time arg and their signatures would be more similar to the lower level methods of the same name.